src/Data/PasswordTokens.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE TemplateHaskell  #-}
+
+{-# OPTIONS_GHC -Wall -Werror #-}
+
+module Data.PasswordTokens
+where
+
+import Control.Lens
+import Data.SafeCopy (base, deriveSafeCopy)
+import Data.Map as Map
+
+import Data.DoubleMap
+import Types
+
+
+-- * type constants
+
+-- See: Data.Delegation for further explanation
+
+type U = AUID User
+
+
+-- * types
+
+newtype Validity = Validity { _validityTimeOut :: Timestamp }
+  deriving (Eq, Ord, Read, Show)
+
+data PasswordTokens = PasswordTokens {
+      _passwordTokens :: Map U (Map PasswordToken Validity)
+    }
+  deriving (Eq, Ord, Read, Show)
+
+makeLenses ''PasswordTokens
+
+isTimedOut :: Timestamp -> Validity -> Bool
+isTimedOut ts (Validity v) = ts >= v
+
+
+emptyPasswordTokens :: PasswordTokens
+emptyPasswordTokens = PasswordTokens Map.empty
+
+-- Add a new token to the tokens which will time out on the given timestamp
+newPasswordToken :: U -> PasswordToken -> Timestamp -> PasswordTokens -> PasswordTokens
+newPasswordToken u t ts (PasswordTokens pt) =
+    PasswordTokens $ insertDoubleMap u t (Validity ts) pt
+
+-- Checks if the token is valid and not expired at the given timestamp
+checkValid :: U -> PasswordToken -> Timestamp -> PasswordTokens -> PasswordTokenState
+checkValid u t ts (PasswordTokens pt) =
+    maybe Invalid (\t' -> if isTimedOut ts t' then TimedOut else Valid) $ lookupDoubleMap u t pt
+
+-- Clears the timed out tokens for the given user
```

Comment:
this is one of those documentation lines that leave out precisely what would be interesting to mention, and instead say everything that can be inferred from the function name and type signature.  (-: